### PR TITLE
Add httpHeaders to forking config

### LINKED ISSUE #1588: Support JsonRpcProvider basic authentication for `hardhat.forking.url`
`networks.hardhat.forking.url` does not accept a  username or password config like in 
https://github.com/ethers-io/ethers.js/issues/159#issuecomment-381359645

```javascript
var urlInfo = {
    url: 'https://foobar.com/eth',
    user: 'username',
    password: 'password'123
};
```

As such we were not able to connect to our Chainstack archival node that requires http basic authentication.

Related code: https://github.com/nomiclabs/hardhat/blob/8277a7219ee25f41495bd6dfde6bd4e308fc84e4/packages/hardhat-core/src/types/config.ts#L71-L75

### PR DESCRIPTION
**Pull request by TimDaub**
- Fixes #1588
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.

I haven't tested this except for running `npm run watch` (0 errors). I'm happy to contribute more by adding tests, but I'm helplessly overwhelmed by the code base's size. Pls guide me thx.

### CODE REVIEW

**Comment by fvictorio** [comment]
@TimDaub the most useful thing to do would be to test this, if you have a node that requires authentication. The way to do that goes something like this:

```
cd packages/hardhat-core
yarn && yarn build
npm link
```

Then in your project do `npm link hardhat`, add the httpHeaders, and see if your changes work.

**Comment by TimDaub** [comment]
```
npx hardhat node
Error HH110: Invalid JSON-RPC response received: <html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></center>
<hr><center>nginx/1.18.0 (Ubuntu)</center>
</body>
</html>


For more info go to https://hardhat.org/HH110 or run Hardhat with --show-stack-traces
```

```js
module.exports = {
  networks: {
    hardhat: {
      forking: {
        url: "https://bla",
        httpHeaders: {
          Authorization: "Bearer token"
        }
      }
    }
  }
};
```

It definitely reaches my node and it rejects it. I wanted to see if in `makeForkClient`, the `forkConfig` contained it. But when I console.log `forkConfig` I get nothing. No logs. So now I'm scratching my head...

**Comment by fvictorio** [comment]
> It definitely reaches my node and it rejects it. I wanted to see if in `makeForkClient`, the `forkConfig` contained it. But when I console.log `forkConfig` I get nothing. No logs. So now I'm scratching my head...

Weird. I'll try to take a look as soon as I have some time.

File: packages/hardhat-core/src/internal/core/providers/construction.ts [hunk 1]
```diff
@@ -69,6 +69,7 @@ export function createProvider(
       forkConfig = {
         jsonRpcUrl: hardhatNetConfig.forking?.url,
         blockNumber: hardhatNetConfig.forking?.blockNumber,
+        httpHeaders: hardhatNetConfig.forking.httpHeaders,
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/core/providers/construction.ts:72]: Ugh, I should've made this into a separate commit, but @TimDaub this was the only thing that was missing from your PR, pretty close! :smile:

**Comment by fvictorio** [comment]
This is ready for review now. While working on this I discovered a very subtle problem with `hardhat_reset`, but I'll just open a new issue for that.

**Comment by TimDaub** [comment]
So

> Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/nomiclabs/hardhat?pullRequest=1791) before we can accept your contribution.
> **1** out of **2** committers have signed the CLA.

Hey sorry, I'm not a lawyer. Why do I have to sign this agreement? Can you please help me understand. Am I giving away the rights to my code here?

**Comment by alcuadrado** [comment]
> Hey sorry, I'm not a lawyer. Why do I have to sign this agreement? Can you please help me understand.

Two reasons mostly: we are relicensing hardhat network to MIT, and transferring the ip to another entity.

> Am I giving away the rights to my code here?

Short answer: yes

We understand this is annoying, but our current licensing situation is a mess, and this is the only way to avoid it.

**Comment by TimDaub** [comment]
What do I get in return for signing this CLA? I did this work mainly to aid in developing "the commons". Now that I'm asked to waive my rights; I don't see my effort being compensated.

**Comment by alcuadrado** [comment]
That's a very valid point, @TimDaub. 

The complete answer is pretty complicated, and I'm not a lawyer either, but I'll make my best here: we are creating a non-profit foundation that's going to own hardhat (soon to be announced). We want to relicense it as mit and donate it to this new foundation. I think the "contributing to the commons" spirit is still valid in this situation. The problem is that we can't do it now, because hardhat network has a custom license which is pretty restrictive. If we merge this PR your contribution will have the same license (this is imposed by Github TOS), and will forbid us to donate it, and technically maybe even use it.

**Comment by TimDaub** [comment]
ping, I believe this is ready for merge? I signed.

**Comment by TimDaub** [comment]
Guys can we finally merge this I've been trying to land this into the main branch since last year August. Just to be extra clear here, my experience contributing code to hardhat wasn't good

- I was asked to sign a contributor agreement that signs away my IP: bad
- @fvictorio helped me a lot with making the change: great
- I waited since November 2021 and nothing has happened + now the branch has conflicts :(

**Comment by fvictorio** [comment]
Hi @TimDaub, I opened a new PR (#2795) to make a small change and will merge it as soon as the CI checks finish.

You are of course absolutely right to be upset, I'm very sorry. We are overloaded with things to do, but that's not excuse. This was a very well-done PR for a useful feature; we should've merged it sooner.

Thanks again for working on this. The change will be included in the next release but, if you want to try it right now, I published it under the `dev` tag. You can install it with `npm i hardhat@dev`. Let me know if you find any issues.

**Pull request merged**